### ISSUE TITLE
[BUG] junos.install_config does not render the template when template_vars is omitted

### ISSUE THREAD

**Issue by bswilden** [saltstack/salt#57665]
**Description**
There’s an issue in https://pypi.org/project/salt/3001rc1/  with the `junos.install_config` function. 

Previously, if you called the `junos.install_config` function with a template that used only pillar or grain data, and you omitted the `template_vars` parameter, it would render the template and apply the configuration. Now it’s giving an error. 

The prior behavior was that the it would render the configuration template in all cases, even if `template_vars` wasn’t provided. Now that behavior has changed and breaks with templates that don’t consume `template_vars`.

**Expected behavior**
When the `template_vars` parameter is omitted for the `junos.install_config` function, the template should still render.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3001rc1

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: Not Installed
         Python: 3.6.9 (default, Apr 18 2020, 01:56:04)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 19.0.1
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2

System Versions:
           dist: ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-99-generic
         system: Linux
        version: Ubuntu 18.04 bionic
```
</details>

**Additional context**
I omitted the error originally, because it doesn't reflect the real issue. Here it is. It's a Junos OS ConfigLoadError because the template doesn't render correctly when `template_vars` is omitted. In the previous release, the example worked correctly - it rendered the template and loaded the configuration on the device.

```
user@host:~$ sudo salt 'r*' state.apply
r1:
----------
          ID: Apply BGP configuration
    Function: junos.install_config
        Name: salt://configs-master/junos-config-bgp-template.set
      Result: False
     Comment:
     Started: 22:37:58.713002
    Duration: 525.159 ms
     Changes:
              ----------
              format:
                  set
              message:
                  Could not load configuration due to : "ConfigLoadError(severity: error, bad_element: {%, message: error: unknown command
                  error: invalid ip address or hostname: {{
                  error: as_num: '{{': Use format 'x' or 'x.y' to specify number 'x' (Range 1 .. 4294967295) or number with higher 2 bytes 'x' (Range 0 .. 65535) and lower 2 bytes 'y' (Range 0 .. 65535) respectively
                  error: syntax error, expecting [Enter]
                  error: invalid ip address or hostname: {{
                  error: unknown command
                  error: invalid interface type: {{
                  error: invalid interface type: {{
                  error: invalid ip address or hostname: {{
                  error: invalid interface type: {{
                  error: statement creation failed
                  error: unknown command
                  error: unknown command)"
              out:
                  False

Summary for r1
------------
Succeeded: 0 (changed=1)
Failed:    1
------------
Total states run:     1
Total run time: 525.159 ms
```

Sample template, which worked fine previously:
```
{% if pillar.BGP_data %}

set interfaces lo0 unit 0 family inet address {{ pillar.BGP_data.loopback }}/32
set policy-options policy-statement bgp-ecmp then load-balance per-packet
set policy-options policy-statement bgp-in then accept
set policy-options policy-statement bgp-out then next-hop self
set policy-options policy-statement bgp-out then accept
set protocols bgp group underlay type external
set protocols bgp group underlay import bgp-in
set protocols bgp group underlay export bgp-out
set protocols bgp group underlay local-as {{ pillar.BGP_data.local_asn }}
set protocols bgp group underlay multipath multiple-as
set routing-options router-id {{ pillar.BGP_data.loopback }}
set routing-options forwarding-table export bgp-ecmp

{% for neighbor in pillar.BGP_data.neighbors %}
set interfaces {{ neighbor.interface }} unit 0 description {{ neighbor.name }}
set interfaces {{ neighbor.interface }} unit 0 family inet address {{ neighbor.local_ip }}/30
set protocols bgp group underlay neighbor {{ neighbor.peer_ip }} peer-as {{ neighbor.asn }}
set protocols lldp interface {{ neighbor.interface }}
{% endfor %}

{% endif %}
```

<details><summary>template sample data</summary>

```
::::::::::::::
r1.sls
::::::::::::::
BGP_data:
  loopback: 192.168.0.1
  local_asn: 64521
  neighbors:
     - interface: ge-0/0/0
       name: to-r2
       asn: 64522
       peer_ip: 198.51.100.2
       local_ip: 198.51.100.1
       peer_loopback: 192.168.0.2
     - interface: ge-0/0/1
       name: to-r3
       asn: 64523
       peer_ip: 198.51.100.6
       local_ip: 198.51.100.5
       peer_loopback: 192.168.0.3
::::::::::::::
r2.sls
::::::::::::::
BGP_data:
  loopback: 192.168.0.2
  local_asn: 64522
  neighbors:
     - interface: ge-0/0/0
       name: to-r1
       asn: 64521
       peer_ip: 198.51.100.1
       local_ip: 198.51.100.2
       peer_loopback: 192.168.0.1
     - interface: ge-0/0/1
       name: to-r3
       asn: 64523
       peer_ip: 198.51.100.10
       local_ip: 198.51.100.9
       peer_loopback: 192.168.0.3
::::::::::::::
r3.sls
::::::::::::::
BGP_data:
  loopback: 192.168.0.3
  local_asn: 64523
  neighbors:
     - interface: ge-0/0/0
       name: to-r1
       asn: 64521
       peer_ip: 198.51.100.5
       local_ip: 198.51.100.6
       peer_loopback: 192.168.0.1
     - interface: ge-0/0/1
       name: to-r2
       asn: 64522
       peer_ip: 198.51.100.9
       local_ip: 198.51.100.10
       peer_loopback: 192.168.0.2
```
</details>

**Comment by sagetherage** [comment]
@bswilden can you please provide the salt version report and at least the text of the error you are seeing, please?

**Comment by bswilden** [comment]
@sagetherage I added the version history and error.

**Comment by dmurphy18** [comment]
@bswilden Do you have sample pillar.BGP_data that can be used including valid neighbor parameters.  It would help in ensuring valid testing of the issue

**Comment by bswilden** [comment]
@sagetherage Test data added

**Comment by dmurphy18** [comment]
@bswilden What previous version of Salt was this working on (versions-report if possible)

I am finding it failing with r1 pillar data on Salt 3000.3
```
root@Unknown:/srv/salt# salt t18 test.versions
t18:
    Salt Version:
               Salt: 3000.3
     
    Dependency Versions:
               cffi: Not Installed
           cherrypy: Not Installed
           dateutil: 2.6.1
          docker-py: Not Installed
              gitdb: 2.0.3
          gitpython: 2.1.8
             Jinja2: 2.10
            libgit2: Not Installed
           M2Crypto: Not Installed
               Mako: 1.0.7
       msgpack-pure: Not Installed
     msgpack-python: 0.5.6
       mysql-python: Not Installed
          pycparser: Not Installed
           pycrypto: 2.6.1
       pycryptodome: Not Installed
             pygit2: Not Installed
             Python: 3.6.9 (default, Apr 18 2020, 01:56:04)
       python-gnupg: 0.4.1
             PyYAML: 3.12
              PyZMQ: 16.0.2
              smmap: 2.0.3
            timelib: Not Installed
            Tornado: 4.5.3
                ZMQ: 4.2.5
     
    System Versions:
               dist: Ubuntu 18.04 bionic
             locale: UTF-8
            machine: x86_64
            release: 4.15.0-101-generic
             system: Linux
            version: Ubuntu 18.04 bionic
     
root@Unknown:/srv/salt# salt t18 state.show_sls test
t18:
    ----------
    apply_bgp_conf:
        ----------
        __env__:
            base
        __sls__:
            test
        junos:
            |_
              ----------
              name:
                  salt://configs-master/junos-config-bgp-template.set
            - install_config
            |_
              ----------
              order:
                  10000
root@Unknown:/srv/salt# echo "" >/var/log/salt/master
root@Unknown:/srv/salt# salt t18 state.sls test
t18:
----------
          ID: apply_bgp_conf
    Function: junos.install_config
        Name: salt://configs-master/junos-config-bgp-template.set
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python3/dist-packages/salt/state.py", line 1981, in call
                  **cdata['kwargs'])
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 1977, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/states/junos.py", line 27, in wrapper
                  ret = function(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/states/junos.py", line 349, in install_config
                  ret['changes'] = __salt__['junos.install_config'](name, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 1248, in __getitem__
                  func = super(LazyLoader, self).__getitem__(item)
                File "/usr/lib/python3/dist-packages/salt/utils/lazy.py", line 108, in __getitem__
                  raise KeyError(key)
              KeyError: 'junos.install_config'
     Started: 15:22:48.636566
    Duration: 3610.021 ms
     Changes:   

Summary for t18
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:   3.610 s
ERROR: Minions returned with non-zero exit code
root@Unknown:/srv/salt# 
root@Unknown:/srv/salt# cat /srv/pillar/bgp.sls 
## ::::::::::::::
## r1.sls
## ::::::::::::::
BGP_data:
  loopback: 192.168.0.1
  local_asn: 64521
  neighbors:
    - interface: ge-0/0/0
      name: to-r2
      asn: 64522
      peer_ip: 198.51.100.2
      local_ip: 198.51.100.1
      peer_loopback: 192.168.0.2
    - interface: ge-0/0/1
      name: to-r3
      asn: 64523
      peer_ip: 198.51.100.6
      local_ip: 198.51.100.5
      peer_loopback: 192.168.0.3
root@Unknown:/srv/salt# cat configs-master/junos-config-bgp-template.set 
{% if pillar.BGP_data %}

set interfaces lo0 unit 0 family inet address {{ pillar.BGP_data.loopback }}/32
set policy-options policy-statement bgp-ecmp then load-balance per-packet
set policy-options policy-statement bgp-in then accept
set policy-options policy-statement bgp-out then next-hop self
set policy-options policy-statement bgp-out then accept
set protocols bgp group underlay type external
set protocols bgp group underlay import bgp-in
set protocols bgp group underlay export bgp-out
set protocols bgp group underlay local-as {{ pillar.BGP_data.local_asn }}
set protocols bgp group underlay multipath multiple-as
set routing-options router-id {{ pillar.BGP_data.loopback }}
set routing-options forwarding-table export bgp-ecmp

{% for neighbor in pillar.BGP_data.neighbors %}
set interfaces {{ neighbor.interface }} unit 0 description {{ neighbor.name }}
set interfaces {{ neighbor.interface }} unit 0 family inet address {{ neighbor.local_ip }}/30
set protocols bgp group underlay neighbor {{ neighbor.peer_ip }} peer-as {{ neighbor.asn }}
set protocols lldp interface {{ neighbor.interface }}
{% endfor %}

{% endif %}
root@Unknown:/srv/salt# 
```
```

**Comment by dmurphy18** [comment]
Similar errors with Salt 2019.2.5

**Comment by bswilden** [comment]
Ok. I thought it was a 2019.2 build, but maybe it was an interim Juniper build based on 2019.2. Sorry Let me check with Juniper then.

**Comment by bswilden** [comment]
For what it's worth, the Juniper Day One: Automating Junos with Salt guide uses the bootstrap script to install the 2018.3.2 Oxygen release and uses the install_config function to load a similar template that uses pillar data in Chapter 6.
https://www.juniper.net/documentation/en_US/day-one-books/DO_Automating_SALT.pdf

So it seems like it worked in that release   I had based my example on that one. I know I tested on some version of 2019.2, but it might have been a Juniper build from GitHub.  However, according to the book, this functionality worked prior to that, so it seems like that feature was added prior to the build I tested on, and was not added as part of that build.

**Comment by bswilden** [comment]
I backed the same setup down to 2019.2.5 (installed from PyPI), and i was able to get the template to load. This setup has a separate Salt master and proxy minion server.  I also tried on a setup with a single server for the salt-master and proxy minions and I was able to load the template in various releases on that setup including 2018.3.2, 2019.2.2, 2019.2.4, 2019.2.5.  I'm including the state file also.

```
Apply BGP configuration:
  junos.install_config:
    - name: salt://configs-master/junos-config-bgp-template.set
```

This is the versions-report for the separate master and proxy minion server setup, in that order:
<details><summary>master / proxy minion versions</summary>

```
Salt Version:
           Salt: 2019.2.5

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: Not Installed
         Python: 3.6.9 (default, Apr 18 2020, 01:56:04)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 19.0.1
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2

System Versions:
           dist: Ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-99-generic
         system: Linux
        version: Ubuntu 18.04 bionic
```

```
Salt Version:
           Salt: 2019.2.5

Dependency Versions:
           cffi: 1.14.0
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.11.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: 2.6.1
   pycryptodome: 3.9.7
         pygit2: Not Installed
         Python: 3.6.9 (default, Apr 18 2020, 01:56:04)
   python-gnupg: 0.4.1
         PyYAML: 3.13
          PyZMQ: 19.0.1
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2

System Versions:
           dist: Ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-99-generic
         system: Linux
        version: Ubuntu 18.04 bionic

```

</details>

**Comment by dmurphy18** [comment]
Ah! proxy minions, being using regular salt-minions, but it is the same code executed on both.  Juniper have been making changes on both the state and execution modules junos.py for Salt 3001 (just gone live). Differences in 2019.2.4 and 2019.2.5 is down to CVE fixes. State file very similar to the one I tested with too. Will try 2018.3.2 on a clean box and work from there.

**Comment by bswilden** [comment]
Sorry if it wasn't clear that proxy minions were in use. As far as I'm aware, the older code only supported using proxy minions for Junos devices, so I didn't think to mention it.

**Comment by dmurphy18** [comment]
My fault, working too much on a native minion for Juniper that works on the actual switch, and forgot there are other ways to use it.

**Comment by dmurphy18** [comment]
When using proxy certain values like keywords are not fully transferred unlike the native minion with a direct connection.

The Junos layers underneath on the device object to extraneous keywords for a number of commands when executed from the CLI, for example: junos.zeroize, junos.lock, junos.unlock, etc.  

The native minion has tests written to catch these, but it is still under development and are not pytest compatible but actual system tests to run form the command line, as such, need rework to fit into testing framework.

However, any fixes need to be cognizant of the native minion and it's direct connection to the Junos layers.

Fix needs to allow for rendering regardless of template_vars present but render all keywords present

**Comment by sagetherage** [comment]
We were unable to get to this in Magnesium moving to Aluminium release

**Comment by dmurphy18** [comment]
@bswilden This appears related to https://github.com/saltstack/salt/issues/59333 in that missing kwargs
Will work on https://github.com/saltstack/salt/issues/59333 since have the issue in debug state and test this once that issue is fixed.

Sorry it has taken a long time to resolve this, other items intervened

**Comment by dmurphy18** [comment]
Moved to Silicon to provide time to write test-cases and fix #59333
